File: backend/app/services/knowledge.py
"""知识页聚合服务：把章节知识点整理成跨章概念图谱。

数据链路（契约与 tools/kb-agent 的 V2 概念层对齐）：

    章节原文锚点 ──► concepts.extract_concepts（LLM 抽取 / 规则回退，带缓存）
                 ──► concepts.build_concept_graph（跨章合并 + 前置关系 + 未解析前置）
                 ──► /api/books/{id}/knowledge

返回结构中 `title/summary/chapterId/chapterTitle/sourceId/status/mastery` 为原有的
兼容字段，`definition/chapters/anchorCount/prerequisites/unresolved` 为 v2 新增字段，
前端据此把「章节内学习顺序」升级为带前置依赖的语义图谱。
"""
from typing import Any, Dict, List

from ..db import Database
from . import concepts as concept_service


def get_knowledge(db: Database, llm, book: Dict[str, Any]) -> Dict[str, Any]:
    """聚合一本教材的知识点图谱。"""
    chapter_blocks: List[Dict[str, Any]] = []
    for chapter in db.chapters_of(book["id"]):
        sections = db.sections_of(book["id"], chapter["id"])
        anchors = db.anchors_of(book["id"], chapter["id"])
        payload = concept_service.extract_concepts(db, llm, chapter, sections, anchors)
        chapter_blocks.append(
            {
                "chapter_id": chapter["id"],
                "chapter_title": chapter["title"],
                "method": payload.get("method", "rule"),
                "concepts": payload.get("concepts", []),
            }
        )

    graph = concept_service.build_concept_graph(
        chapter_blocks, db.progress_of_book(book["id"])
    )
    return {
        "bookId": book["id"],
        "bookTitle": book["title"],
        "concepts": graph["concepts"],
        "relations": graph["relations"],
        "unresolved": graph["unresolved"],
        "methods": graph["methods"],
        "stats": graph["stats"],
    }

File: frontend/src/data/knowledge.js
import { books, studyContents } from './books.js';

/**
 * 演示模式的语义关系标注：知识点 id → 它依赖的前置概念名。
 *
 * 真实数据由后端知识抽取生成（见 backend/app/services/concepts.py）；
 * 这里手工标注，保证「无模型、无后端」时演示模式与后端模式看到同一套字段。
 * 其中「极限」在演示数据里没有对应的知识点卡片，用于演示「教材外前置」。
 */
const DEMO_PREREQUISITES = {
  'kp-rate': ['极限'],
  'kp-limit': ['先抓住「变化率」'],
};

/** 演示模式使用的知识点图谱数据（结构与后端 /api/books/{id}/knowledge 一致）。 */
export function knowledgeFor(bookId, progressByChapter = {}) {
  const book = books.find((item) => item.id === bookId);
  const concepts = [];
  const byTitle = new Map();

  for (const chapter of book?.chapters ?? []) {
    const content = studyContents[bookId]?.[chapter.id];
    for (const point of content?.knowledgePoints ?? []) {
      const summary = point.body.map((part) => part.v).join('');
      const sourceId = point.sourceId ?? null;
      const concept = {
        id: `${chapter.id}:${point.id}`,
        title: point.title,
        label: point.title,
        definition: summary,
        summary,
        chapterId: chapter.id,
        chapterTitle: content.heading,
        chapters: [chapter.id],
        sourceId,
        sourceLabel: point.sourceLabel ?? '定位教材：原文',
        anchors: sourceId ? [sourceId] : [],
        anchorCount: sourceId ? 1 : 0,
        prerequisites: [...(DEMO_PREREQUISITES[point.id] ?? [])],
        method: 'demo',
        status: progressByChapter[chapter.id]?.status ?? 'planned',
        mastery: progressByChapter[chapter.id]?.mastery ?? 0,
      };
      concepts.push(concept);
      byTitle.set(concept.title, concept);
    }
  }

  const relations = [];
  const linked = new Set();
  const external = new Map();

  for (const concept of concepts) {
    for (const name of concept.prerequisites) {
      const target = byTitle.get(name);
      if (target) {
        linked.add(`${concept.id}->${target.id}`);
        relations.push({
          id: `${concept.id}->${target.id}`,
          source: concept.id,
          target: target.id,
          type: 'prerequisite',
          label: '前置',
        });
        continue;
      }
      if (!external.has(name)) external.set(name, new Set());
      external.get(name).add(concept.title);
    }
  }

  // 学习顺序兜底边：同章节相邻知识点，若已有前置边则不再重复连接
  for (let i = 1; i < concepts.length; i += 1) {
    const previous = concepts[i - 1];
    const current = concepts[i];
    if (linked.has(`${previous.id}->${current.id}`) || linked.has(`${current.id}->${previous.id}`)) {
      continue;
    }
    linked.add(`${previous.id}->${current.id}`);
    relations.push({
      id: `${previous.id}->${current.id}`,
      source: previous.id,
      target: current.id,
      type: 'sequence',
      label: '学习顺序',
    });
  }

  const unresolved = [...external.entries()]
    .sort(([a], [b]) => a.localeCompare(b))
    .map(([name, requiredBy]) => ({ name, requiredBy: [...requiredBy] }));

  return {
    bookId,
    bookTitle: book?.title ?? '',
    concepts,
    relations,
    unresolved,
    methods: ['demo'],
    stats: {
      conceptCount: concepts.length,
      learnedCount: concepts.filter((concept) => concept.status !== 'planned').length,
      relationCount: relations.length,
      prerequisiteCount: relations.filter((relation) => relation.type === 'prerequisite').length,
      unresolvedCount: unresolved.length,
    },
  };
}
